source/blog/2015-05-08-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,216 @@
+---
+title: How RSpec 3.0 composable matchers work
+author: Sam Phippen
+---
+
+##Warmup: the `===` protocol
+
+In Ruby, a number of objects implement the `===` protocol. Some of the objects
+that implement this protocol include:
+
+* all class objects
+* lambdas
+* regexes
+* ranges
+
+Conceptually, this protocol represents a "match". That is, the object that
+implements the `===` declares whether or not it "matches" the other object.
+For example, `Array === []` returns `true`. This is because an empty array is a
+subclass of the array class. `Array === {}` returns false because a hash is not
+a subclass of array. For regexes, `===` returns true if the passed string
+matches the regex, and false if it does not.
+
+The `===` method is used in Ruby case statements to decide if a branch should
+be taken. For example
+
+```
+case 3
+when (1...18)
+    puts "hello"
+end
+```
+
+will print hello, because `(1...18) === 3` returns `true`.
+
+The new RSpec matchers make heavy use of `===` internally, if you'd like to get
+more of an understanding of `===`, I'd heavily suggest you pop open IRB and
+play around with it a little before moving on.
+
+## RSpec matchers
+
+In an RSpec expectation expression like `expect(object_1).to eq(object_2)`
+there are two primary objects at play, the **target** and the **matcher**. The
+**target** object is created by "expect" and basically wraps the passed object
+to it in a class called `ExpectationTarget`. The primary methods on the
+`ExpectationTarget` that users see are `to` and `not_to`/`to_not`. Let's focus
+on `to`.
+
+The `to` method is basically responsible for invoking the `matches?` method on the
+passed matcher. It's possible to see this happening quite easily in an IRB session.
+
+```
+>> require "rspec/expectations"
+=> true
+>> o = Object.new
+=> #<Object:0x007fa20d221490>
+>> o.extend(RSpec::Matchers)
+=> #<Object:0x007fa20d221490>
+>> o.expect(3).to Object.new
```

Comment:
I think it's confusing to show an expectation in this form, because all normal usage of expectations use bare `expect` rather than `some_object.expect`.  Rather than doing `o = Object.new; o.extend(RSpec::Matchers); o.expect(...)` you can do `include RSpec::Matchers; expect(...)`, which mirrors how users use the API.